Core/HLE/sceHttp.cpp

```diff
@@ -15,68 +15,362 @@
+	int err = client.SendRequestWithData(methodstr.c_str(), req, std::string(postData ? postData : "", postData ? postDataSize : 0), extraHeaders.c_str(), &progress_);
+	if (cancelled_) {
+		return SCE_HTTP_ERROR_ABORTED;
+	}
+	if (err < 0) {
+		return err; // SCE_HTTP_ERROR_BAD_RESPONSE;
+	}
+
+	// Retrieve Response's Status Code (and Headers too?)
+	responseCode_ = client.ReadResponseHeaders(&buffer_, responseHeaders_, &progress_, &httpLine_);
+	if (cancelled_) {
+		return SCE_HTTP_ERROR_ABORTED;
+	}
+
+	// TODO: Read response entity within readData() in smaller chunk(based on size arg of sceHttpReadData) instead of the whole content at once here
+	net::Buffer entity_;
+	int res = client.ReadResponseEntity(&buffer_, responseHeaders_, &entity_, &progress_);
+	if (res != 0) {
+		ERROR_LOG(SCENET, "Unable to read HTTP response entity: %d", res);
+	}
+	entity_.TakeAll(&responseContent_);
+	if (cancelled_) {
+		return SCE_HTTP_ERROR_ABORTED;
+	}
+
+	return 0;
+}
+
+static void __HttpNotifyLifecycle(CoreLifecycle stage) {
+	if (stage == CoreLifecycle::STOPPING) {
+		for (const auto& it : httpObjects) {
+			if (it->className() == name_HTTPRequest)
+				(static_cast<HTTPRequest*>(it.get()))->abortRequest();
+		}
+	}
+}
+
+static void __HttpRequestStop() {
+	// This can happen from a separate thread.
+	std::lock_guard<std::mutex> guard(httpLock);
+	for (const auto& it : httpObjects) {
+		if (it->className() == name_HTTPRequest)
+			(static_cast<HTTPRequest*>(it.get()))->abortRequest();
+	}
+}
+
+void __HttpInit() {
+	Core_ListenLifecycle(&__HttpNotifyLifecycle);
+	Core_ListenStopRequest(&__HttpRequestStop);
+
+	// Finding memory leaks by object numbers
+	//_CrtSetDbgFlag(_CRTDBG_ALLOC_MEM_DF | _CRTDBG_LEAK_CHECK_DF);
+	//_CrtSetBreakAlloc(2407638);
+	//_CrtSetBreakAlloc(1691236);
+	//_CrtSetBreakAlloc(1691235);
+	//_CrtSetBreakAlloc(1691231);
+}
+
+void __HttpShutdown() {
+	std::lock_guard<std::mutex> guard(httpLock);
+	httpObjects.clear();
+}
+
+// id: ID of the template or connection
+int sceHttpSetResolveRetry(int id, int retryCount) {
+	WARN_LOG(SCENET, "UNTESTED sceHttpSetResolveRetry(%d, %d)", id, retryCount);
+	if (id <= 0 || id > httpObjects.size())
+		return hleLogError(SCENET, SCE_HTTP_ERROR_INVALID_ID, "invalid id");
+
+	const auto& conn = httpObjects[id - 1LL];
+	if (!(conn->className() == name_HTTPTemplate || conn->className() == name_HTTPConnection))
+		return hleLogError(SCENET, SCE_HTTP_ERROR_INVALID_ID, "invalid id (%s)", conn->className());
+
+	conn->setResolveRetry(retryCount);
+	return 0;
+}
+
+static int sceHttpInit(int poolSize) {
+	WARN_LOG(SCENET, "UNTESTED sceHttpInit(%i)", poolSize);
+	if (httpInited)
+		return hleLogError(SCENET, SCE_HTTP_ERROR_ALREADY_INITED, "http already inited");
+
+	std::lock_guard<std::mutex> guard(httpLock);
+	httpObjects.clear();
+	// Reserve at least 1 element to prevent ::begin() from returning null when no element has been added yet
```

Comment:
That's not really what reserve does?  You'd need to `resize(1)` for that.  `reserve` just makes sure it's ready to have that much memory.  I don't know if an empty container is actually required to have a valid pointer return for `[0]`.

-[Unknown]

Reply:
resize(1) will make count resulting to 1 isn't? 
i only need to allocate the space without increase the count so begin() can return an iterator object, without it apparently returned null instead of an iterator object. 
The reason it must not be null is because on some part of the code the iterator object being increased using +1 i think (i haven't re-read the whole code while fixing the conflict), thus could cause access violation. 

And as i remembered it doesn't have this issue when the httpObjects is freshly created (ie. begin() of an empty vector returns an iteration object as in most c++ examples), only happened when reused i think (or was it the opposite, i kinda forgot😅 was getting a crash because of that null), may be i erased it the wrong way🤔

Reply:
Hm, we should really should have a bool then or something.  Trying to rely on whether there's memory at `&x[0]` when empty just seems like asking for trouble.

-[Unknown]